transport/src/main/java/io/netty/channel/PendingWriteQueue.java

```diff
@@ -37,22 +38,69 @@
             SystemPropertyUtil.getInt("io.netty.transport.pendingWriteSizeOverhead", 64);
 
     private final ChannelHandlerContext ctx;
-    private final ChannelOutboundBuffer buffer;
-    private final MessageSizeEstimator.Handle estimatorHandle;
+    private final PendingTracker tracker;
 
     // head and tail pointers for the linked-list structure. If empty head and tail are null.
     private PendingWrite head;
     private PendingWrite tail;
     private int size;
     private long bytes;
 
+    private interface PendingTracker extends MessageSizeEstimator.Handle {
+        void incrementPendingOutboundBytes(long bytes);
+        void decrementPendingOutboundBytes(long bytes);
+    }
+
     public PendingWriteQueue(ChannelHandlerContext ctx) {
-        if (ctx == null) {
-            throw new NullPointerException("ctx");
+        this.ctx = ObjectUtil.checkNotNull(ctx, "ctx");
+        if (ctx.pipeline() instanceof DefaultChannelPipeline) {
+            final DefaultChannelPipeline pipeline = (DefaultChannelPipeline) ctx.pipeline();
+            tracker = new PendingTracker() {
+                @Override
+                public void incrementPendingOutboundBytes(long bytes) {
+                    pipeline.incrementPendingOutboundBytes(bytes);
+                }
+
+                @Override
+                public void decrementPendingOutboundBytes(long bytes) {
+                    pipeline.decrementPendingOutboundBytes(bytes);
+                }
+
+                @Override
+                public int size(Object msg) {
+                    return pipeline.estimatorHandle().size(msg);
+                }
+            };
+        } else {
+            final ChannelOutboundBuffer buffer = ctx.channel().unsafe().outboundBuffer();
+            final MessageSizeEstimator.Handle estimator = ctx.channel().config().getMessageSizeEstimator().newHandle();
+            tracker = new PendingTracker() {
+                @Override
+                public void incrementPendingOutboundBytes(long bytes) {
+                    // We need to guard against null as channel.unsafe().outboundBuffer() may returned null
+                    // if the channel was already closed when constructing the PendingWriteQueue.
+                    // See https://github.com/netty/netty/issues/3967
+                    if (buffer != null) {
```

Comment:
can you just create a no-op `PendingTracker` if `buffer == null`? The value of `buffer` won't change after its assigned.